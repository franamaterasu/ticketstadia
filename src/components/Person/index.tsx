import { useDispatch, useSelector } from 'react-redux';
import { Profile } from '../../types';
import { useLocation } from 'react-router-dom';
import { addFriend, deleteFriend } from '../../store/reducers/friendsSlice';

type PersonProps = {
  info: Profile;
};

const Person = ({ info }: PersonProps) => {
  const { friends } = useSelector((state) => state.friends);

  const friendExist = friends.some(
    (existFriend: Profile) => existFriend.id === info.id
  );

  const { imagen, nombre, email } = info;

  const dispatch = useDispatch();

  const handleAddClick = (friend: Profile) => {
    dispatch(addFriend(friend));
  };

  const handleDeleteClick = (id: number) => {
    dispatch(deleteFriend(id));
  };

  const location = useLocation();

  return (
    <div className='w-full'>
      <div className='bg-slate-500 py-5 rounded-t-lg'>
        <img
          className='w-24 h-24 mx-auto rounded-full boder-t'
          src={imagen}
          alt={nombre}
        />
      </div>
      <div className='text-center bg-white py-10 rounded-b-lg'>
        <h3 className='text-xl font-semibold'>{nombre}</h3>
        <p className='font-light mb-5'>{email}</p>

        {location.pathname === '/' && (
          <button
            className='bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded disabled:bg-blue-300'
            disabled={friendExist}
            onClick={() => handleAddClick(info)}>
            {friendExist ? 'Amigo agregado' : 'Agregar amigo'}
          </button>
        )}

        {location.pathname !== '/' && (
          <button
            className='bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded'
            onClick={() => handleDeleteClick(info.id)}>
            Eliminar amigo
          </button>
        )}
      </div>
    </div>
  );
};

export default Person;
