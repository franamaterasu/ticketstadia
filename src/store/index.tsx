import { configureStore } from '@reduxjs/toolkit';
import eventsReducer from './reducers/eventsSlice';
import friendsReducer from './reducers/friendsSlice';
import cartReducer from './reducers/cartSlice';

export default configureStore({
  reducer: {
    events: eventsReducer,
    friends: friendsReducer,
    cart: cartReducer,
  },
});
