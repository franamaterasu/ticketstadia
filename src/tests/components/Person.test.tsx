import { describe, test, expect } from 'vitest';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import user from '@testing-library/user-event';
import Person from '../../components/Person';
import store from '../../store';
import { Provider } from 'react-redux';

type WrapperProps = {
  children: JSX.Element;
};

const Wrapper = ({ children }: WrapperProps) => {
  return <Provider store={store}>{children}</Provider>;
};

const PersonMock = {
  id: 3,
  imagen: 'https://randomuser.me/api/portraits/men/61.jpg',
  nombre: 'Lucas Gordon',
  ciudad: 'Barcelona',
  telefono: '650344452',
  email: 'lucas.gordon@example.com',
};

describe('<Person/>', () => {
  test('should match with snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <Wrapper>
          <Person info={PersonMock} />
        </Wrapper>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test('should render information about one Person', () => {
    render(
      <MemoryRouter>
        <Wrapper>
          <Person info={PersonMock} />
        </Wrapper>
      </MemoryRouter>
    );

    const image = screen.getByRole('img', {
      name: 'Lucas Gordon',
    });

    const name = screen.getByRole('heading', {
      level: 3,
      name: 'Lucas Gordon',
    });

    const email = screen.getByText('lucas.gordon@example.com');

    const button = screen.getByRole('button', {
      name: 'Agregar amigo',
    });

    expect(image).toBeDefined();
    expect(name).toBeDefined();
    expect(email).toBeDefined();
    expect(button).toBeDefined();
  });

  test('should render button disabled and text changed when click on it', async () => {
    user.setup();

    render(
      <MemoryRouter>
        <Wrapper>
          <Person info={PersonMock} />
        </Wrapper>
      </MemoryRouter>
    );

    const initialButton = screen.getByRole('button', {
      name: 'Agregar amigo',
    });

    await user.click(initialButton);

    const clickedButton = screen.getByRole('button', {
      name: 'Amigo agregado',
    });

    expect(clickedButton).toBeDefined();

    expect(clickedButton).toBeDisabled();
  });
});
