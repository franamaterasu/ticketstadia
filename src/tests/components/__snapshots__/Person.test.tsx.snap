// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Person/> > should match with snapshot 1`] = `
<div>
  <div
    class="w-full"
  >
    <div
      class="bg-slate-500 py-5 rounded-t-lg"
    >
      <img
        alt="Lucas Gordon"
        class="w-24 h-24 mx-auto rounded-full boder-t"
        src="https://randomuser.me/api/portraits/men/61.jpg"
      />
    </div>
    <div
      class="text-center bg-white py-10 rounded-b-lg"
    >
      <h3
        class="text-xl font-semibold"
      >
        Lucas Gordon
      </h3>
      <p
        class="font-light mb-5"
      >
        lucas.gordon@example.com
      </p>
      <button
        class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded disabled:bg-blue-300"
      >
        Agregar amigo
      </button>
    </div>
  </div>
</div>
`;
